Add WCCOM Site REST API

gedex opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?

### Changes proposed in this Pull Request:

Adds internal REST API to allow WooCommerce.com instructing connected store to install purchased products.

* Namespace for this API: `/wccom-site/`.
* Available routes for WCCOM Site Installer:
    * `GET /wccom-site/v1/installer` to retrieve product installation state.
    * `POST /wccom-site/v1/installer` to initiate product installation.
    * `DELETE /wccom-site/v1/installer` to reset product installation state.
* Request is authenticated using `access_token` and `access_token_secret` that are stored in site option after connecting to WooCommerce.com  (via **WooCommerce > Extensions > WooCommerce.com Subscriptions**).

### How to test the changes in this Pull Request:

To test this, your WooCommerce site needs to be connected to WooCommerce.com (via **WooCommerce > Extensions > WooCommerce.com Subscriptions**). Once connected, you will need to generate signature for the request. For example if your WooCommerce site is https://example.com, signature for `GET /wccom-site/v1/installer` can be generated with:

```php
require_once( WC_ABSPATH . 'includes/admin/helper/class-wc-helper.php' );

$auth = WC_Helper_Options::get( 'auth' );
$access_token_secret = $auth['access_token_secret'];
$signature = hash_hmac(
	'sha256',
	json_endode( [
		'host' => 'example.com',
		'request_uri' => '/wp-json/wccom-site/v1/installer',
		'method' => 'GET',
	] ),
	$access_token_secret
);
```

and then pass `$signature` as `X-Woo-Signature` header.  For `POST /wccom-site/v1/installer` , body needs to be included in:

```php
hash_hmac(
	'sha256',
	json_endode( [
		'host' => 'example.com',
		'request_uri' => '/wp-json/wccom-site/v1/installer',
		'method' => 'GET',
		'body' => $body,
	] ),
	$access_token_secret
);
```

In addition to `X-Woo-Signature`, bearer token (`$access_token`) must be set in the `Authorization` header.

#### Example Requests and Responses

**Retrieve installation state**

```
curl -X GET 'https://example.com/wp-json/wccom-site/v1/installer' \
	-H 'Authorization: Bearer c273e7987e92def6d1e9ce0a38d3671f377cfd56' \
	-H 'X-Woo-Signature: a27dc138d33fd08408e8f9997ad17d78214223019bf3d8d929c212c19265a3e0'

HTTP/1.1 200 OK
{
  "status": "has_error",
  "steps": {
    "18627": {
      "download_link": "",
      "product_type": "plugin",
      "last_step": "move_product",
      "last_error": "Destination folder already exists.",
      "download_path": "/tmp/woocommerce-gateway-stripe.4.2.2-MVTk0q.tmp",
      "unpacked_path": "/var/www/html/wp-content/upgrade/woocommerce-gateway-stripe.4.2.2-MVTk0q",
      "installed_path": "",
      "activate": true,
      "download_url": "https://downloads.wordpress.org/plugin/woocommerce-gateway-stripe.4.2.2.zip"
    }
  },
  "current_step": {
    "product_id": 18627,
    "step": "move_product"
  }
}
```

**Install products**

```
curl -X POST 'https://example.com/wp-json/wccom-site/v1/installer' \
	-H 'Authorization: Bearer c273e7987e92def6d1e9ce0a38d3671f377cfd56' \
	-H 'X-Woo-Signature: a25fa349425ff5fa6adf17b9b5a67cc6f2577b73813757c419a3b9638243a86d' \
	-d '{"products": {"18627": {"activate": true}}}'

HTTP/1.1 200 OK
{
  "status": "in-progress",
  "steps": {
    "18627": {
      "download_link": "",
      "product_type": "",
      "last_step": "",
      "last_error": "",
      "download_path": "",
      "unpacked_path": "",
      "installed_path": "",
      "activate": false
    }
  },
  "current_step": null
}
```

`18627` is the product ID for Stripe. Only valid WooCommerce.com product ID can be passed in `products` parameter. For WooCommerce.com products that are not in wporg, valid subscription is required.

**Reset installation state**

```
curl -i -X DELETE 'https://example.com/wp-json/wccom-site/v1/installer' \
	-H 'Authorization: Bearer c273e7987e92def6d1e9ce0a38d3671f377cfd56' \
	-H 'X-Woo-Signature: 2396c50507b8961e69059dcb1966224c94e05b0d61f848da1a4db7060dfde392'

HTTP/1.1 204 No Content
```

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Add WCCOM Site Installer REST API

gedex committed:
Initial work of adding product install API on top of helper.

gedex committed:
Add authentication for helper request.

gedex committed:
Add reset state.

gedex committed:
Include helper product install class in main helper class.

gedex committed:
Rename filename for helper REST API controller.

gedex committed:
Use WP_REST_Server::get_raw_data to get body from the request.

gedex committed:
Move subscription check as part of installation steps.

gedex committed:
Fix typo in error message.

gedex committed:
Add missing description in docblock.

gedex committed:
Make sure to init the upgrader and clean plugins cache before activate.

gedex committed:
Update the state when installation finished.

gedex committed:
Add product info retrieval step before downloading product.

tgglv committed:
Fix PHP CodeSniffer issues

tgglv committed:
Fix typo

tgglv committed:
Add @return types into annotations

tgglv committed:
Store product's `installed_path`

tgglv committed:
Remove redundant line

tgglv committed:
Add WP.org plugin support into Product Activation

tgglv committed:
Make WP.org plugin support installation method names more precise

gedex committed:
Add support for theme installation.

gedex committed:
Make sure to flush cache before checking wccom updates.

gedex committed:
Fix minor wording issue in error message.

gedex committed:
Rename helper to wccom-site.

gedex committed:
Add docblock in controller class.

gedex committed:
Move WCCOM Site related files to `/wccom-site` directory.

Also use queue when installing products so API can respond immediately.

gedex committed:
Adjust products parameter so it can pass installation args.

gedex committed:
Reset current_step state when installing.

kloon commented:
@gedex since the REST API has been moved out to it's own repo this PR will probably need some refactoring. Could you please have a look at what was done in #23957 

Perhaps it might be worth developing this functionality in its own repo and then also include it as part of the package autoloader? Just a suggestion, but if time is limited then we can keep this in the core repo for now, just needs refactoring to work with the new API as a package changes.

tgglv commented on includes/class-woocommerce.php:
```diff
@@ -435,11 +435,17 @@ public function includes() {
 		 */
 		include_once WC_ABSPATH . 'includes/libraries/action-scheduler/action-scheduler.php';
 
+
```
@gedex 
Nitpick: the new line seems redundant.

gedex replied:
Removed in https://github.com/woocommerce/woocommerce/pull/24123/commits/e1743bfadb251bd616375b372102c179911aacdb.

gedex commented:
> REST API has been moved out to it's own repo this PR will probably need some refactoring. Could you please have a look at what was done in #23957

Seems the merge on 23957 is what causing conflicting files in this PR. Will take a look and resolve them.

> Perhaps it might be worth developing this functionality in its own repo and then also include it as part of the package autoloader?

I was wondering about this too and chatted @mikejolley on Slack. I think we agreed to have this lives in core with main reasoning of not a public-facing REST API and how it tightly-coupled with Helper, but tagging you here, Mike, in case we need to develop it in a similar fashion.

gedex committed:
Register wccom-site rest in main class.

With API moved to its own package (see #23957), controllers for wccom-
site namespace are now registered in wccom-site main class.

gedex committed:
Remove extra new line.

gedex committed:
Fix coding standard issues.

gedex commented:
> Seems the merge on 23957 is what causing conflicting files in this PR. Will take a look and resolve them.

Rebased with `master`. See https://github.com/woocommerce/woocommerce/pull/24123/commits/40201eb9c301a71662024de0e29cc27e523e0764 for how `wccom-site/` namespace is registered.

claudiosanches commented on includes/class-wc-rest-authentication.php:
```diff
@@ -66,6 +67,99 @@ protected function is_request_to_rest_api() {
+	 * @return bool
+	 */
+	protected function verify_wccom_request( $body, $signature, $access_token_secret ) {
+		// phpcs:disable
```
Please, don't disable all rules, better to only disable a few ones when required.

gedex replied:
Updated in https://github.com/woocommerce/woocommerce/pull/24123/commits/59a5c6d7545ffc2e1553e2bfbf8376e05eb7f00f.

kloon commented on includes/admin/helper/class-wc-helper.php:
```diff
@@ -1098,6 +1098,18 @@ private static function _get_local_from_product_id( $product_id ) {
+	 * Checks whether current site has product subscription of a given ID.
+	 *
+	 * @param int $product_id The product id.
+	 *
```
Could you please add a @ since tag here?

gedex replied:
Added in https://github.com/woocommerce/woocommerce/pull/24123/commits/f51265e58db0a6eb06d694bb4f1766c6890a6d78.

kloon commented on includes/class-wc-rest-authentication.php:
```diff
@@ -66,6 +67,99 @@ protected function is_request_to_rest_api() {
+	 * @return bool
+	 */
+	protected function is_request_to_wccom_site_rest_api() {
+		if ( empty( $_SERVER['REQUEST_URI'] ) ) {
```
Can we try and avoid using $_SERVER vars for getting the current url, you can get this with `add_query_arg( array() )`

gedex replied:
Updated in https://github.com/woocommerce/woocommerce/pull/24123/commits/284239ee4485548486390feba1863cad51bed9cf.

kloon commented on includes/class-wc-rest-authentication.php:
```diff
@@ -66,6 +67,99 @@ protected function is_request_to_rest_api() {
+		}
+
+		$rest_prefix = trailingslashit( rest_get_url_prefix() );
+		$request_uri = esc_url_raw( wp_unslash( $_SERVER['REQUEST_URI'] ) );
```
Same as above, try and avoid accessing $_SERVER vars for building urls, use `add_query_arg( array() )` instead.

gedex replied:
Updated in https://github.com/woocommerce/woocommerce/pull/24123/commits/284239ee4485548486390feba1863cad51bed9cf.

kloon reviewed (changes requested):
Please see feedback

gedex committed:
Disable few phpcs rules instead of all.

gedex committed:
Add @since tag.

gedex committed:
Replace SERVER['REQUEST_URI'] in favor of add_query_arg.

gedex committed:
Add @ since tag.

gedex committed:
Remove extra docblock line.

claudiosanches commented on includes/class-wc-rest-authentication.php:
```diff
@@ -38,6 +38,7 @@ class WC_REST_Authentication {
 	 * Initialize authentication actions.
 	 */
 	public function __construct() {
+		add_filter( 'determine_current_user', array( $this, 'authenticate_wccom' ), 14 );
```
This should be moved into a new class, since one day we should stop support for all our current authenticate methods.

dwainm replied:
Agree with Claudio, this functionality refers mainly to the helper and should reside alongside othe helper functionality. This also ensures that a helper update doesn't look past the code here.

gedex replied:
Moved in https://github.com/woocommerce/woocommerce/pull/24123/commits/0ffd4db6302a4d14e632ae0e365245b2c496ff67.

claudiosanches reviewed (changes requested):
There's a few things to change

claudiosanches commented on includes/class-wc-rest-authentication.php:
```diff
@@ -66,6 +67,102 @@ protected function is_request_to_rest_api() {
+			return $user_id;
+		}
+
+		if ( empty( $_SERVER['HTTP_AUTHORIZATION'] ) ) {
```
This is not present in `$_SERVER` in all servers, that's why we have a `$this->get_authorization_header()` method, so needs to replace it.

gedex replied:
Added helper method from `get_authorization_header()` in https://github.com/woocommerce/woocommerce/pull/24123/commits/c8812a205bfe249fa6ddb9ab7a209425de73d081. It's copied as is from `class-wc-rest-authentication.php` instead of depending on the class.

claudiosanches commented on includes/class-wc-rest-authentication.php:
```diff
@@ -66,6 +67,102 @@ protected function is_request_to_rest_api() {
+		require_once WC_ABSPATH . 'includes/admin/helper/class-wc-helper-options.php';
+		$access_token = trim( substr( $request_auth, 7 ) );
+		$site_auth    = WC_Helper_Options::get( 'auth' );
+		if ( empty( $site_auth['access_token'] ) || $access_token !== $site_auth['access_token'] ) {
```
Seems like there's an error in here, I can't authenticate properly in this step.

`trim( substr( $request_auth, 7 ) );` is returning the same value that `$site_auth['access_token_secret']` instead of `$site_auth['access_token']`.

Also should use `! hash_equals()` instead of `!==`.

gedex replied:
Fixed in https://github.com/woocommerce/woocommerce/pull/24123/commits/47db7588d70472846e33525a20adff5e8a698fc3.

dwainm commented on includes/wccom-site/class-wc-wccom-site-installer.php:
```diff
@@ -0,0 +1,543 @@
+<?php
```
I'm interested in the reason for not making this part of the helper? Though it is a new feature, we still view this as the same context: Our connection to our merchants, and along with that the ability to "push" *things( any valid payload) to them given that we have a connection.

gedex replied:
This mostly came from team discussion (couldn't recall whether on dev roundtable or demo day, /cc @haszari @mattyza) that in general, we should avoid using term `helper` and be specific with the naming. Of course, it's tightly coupled with helper as it depends on the helper but not vice versa.

dwainm commented on includes/wccom-site/class-wc-wccom-site-installer.php:
```diff
@@ -0,0 +1,543 @@
+		$filename = false;
+
+		// If product is WP.org one, find out its filename.
+		$dir_name = self::get_wporg_product_dir_name( $product_id );
```
For the future, something WCCOM could provide the client? This refers if it is hosted on wp.org or wc.com.

dwainm replied:
Could also simplify function by handling the file retrieval for wp.org and wc.com in a separate function.

dwainm replied:
Same applies to themes.

gedex replied:
> For the future, something WCCOM could provide the client?

We don't store main filename of extension in the product.

> Could also simplify function by handling the file retrieval for wp.org and wc.com in a separate function.

Given a product ID, for WCCOM products there's no special handling as helper caches plugin/theme info with valid subscriptions. However, it's not the case with wporg product in which WCCOM doesn't create subscription, thus no lookup provided.

dwainm commented on includes/wccom-site/rest-api/v1/class-wc-rest-wccom-site-installer-v1-controller.php:
```diff
@@ -0,0 +1,167 @@
+ * @since   3.7.0
+ */
+
+defined( 'ABSPATH' ) || exit;
```
This style should be used with all new files.

gedex replied:
Yep. All new files have that.

dwainm reviewed (approved):
The feedback is not blocking though nice to have. 

I couldn't test as I got this error and didn't understand the issue here:
`Fatal error: Uncaught Error: Class 'WooCommerce' `

I looked at the code and I'm happy with the specific implementation.

When I get a way to fix the error I can do the actual functionality testing.

claudiosanches commented on includes/wccom-site/class-wc-wccom-site.php:
```diff
@@ -0,0 +1,62 @@
+	 */
+	protected static function includes() {
+		require_once WC_ABSPATH . 'includes/admin/helper/class-wc-helper.php';
+		require_once __DIR__ . '/class-wc-wccom-site-installer.php';
```
@kloon is fine to use PHP 5.3 magic constant in here?

gedex replied:
Happy to switch to `WC_ABSPATH`, if you prefer. `__DIR__` is shorter, no strong argument.

gedex replied:
Replaced with `WC_ABSPATH` in https://github.com/woocommerce/woocommerce/pull/24123/commits/3f9bbf5a3bed66349b1af2e0d1ff430abe63a1ec.

claudiosanches commented on includes/wccom-site/rest-api/v1/class-wc-rest-wccom-site-installer-v1-controller.php:
```diff
@@ -0,0 +1,167 @@
+ * @package WooCommerce/WCCOM_Site/REST_API
+ * @extends WC_REST_Controller
+ */
+class WC_REST_WCCOM_Site_Installer_V1_Controller extends WC_REST_Controller {
```
Since is a private API, we probably don't need to use `V1`.

gedex committed:
Remove duplicate state download_link in favor of download_url.

gedex commented on includes/wccom-site/class-wc-wccom-site.php:
```diff
@@ -0,0 +1,62 @@
+		self::includes();
+
+		add_action( 'woocommerce_wccom_install_products', array( 'WC_WCCOM_Site_Installer', 'install' ) );
+		add_action( 'woocommerce_rest_api_get_rest_namespaces', array( __CLASS__, 'register_rest_namespace' ) );
```
@claudiosanches https://github.com/woocommerce/woocommerce-rest-api/blob/46453550e8d7468f6a1f46b47fefd2fdc482c801/src/Server.php#L52-L59

gedex committed:
Move authentication to WCCOM Site API to wccom-site main class.

gedex committed:
Use hash_equals when checking access_token.

gedex committed:
Add helper method get_authorization_header.

This to support scenario in which HTTP_AUTHORIZATION being stripped
and use getallheaders as a fallback. This helper method is copied from
class-wc-rest-authentication.php.

claudiosanches commented on includes/wccom-site/class-wc-wccom-site-installer.php:
```diff
@@ -0,0 +1,543 @@
+	 */
+	private static function get_wporg_plugin_main_file( $dir ) {
+		// Ensure that exact dir name is used.
+		$dir .= '/';
```
Better use `trailingslashit()`.

claudiosanches committed:
Use trailingslashit

claudiosanches committed:
Fixed standards

gedex committed:
Replace __DIR__ with WC_ABSPATH.

claudiosanches committed:
Fixed doc blocks and coding standards

claudiosanches committed:
Merge branch 'add/wccom-site-rest-api' of github.com:woocommerce/woocommerce into add/wccom-site-rest-api

claudiosanches committed:
Better naming for REST API endpoints

claudiosanches commented:
@dwainm Maybe the fatal error that you are seeing is coming from a plugin, since I can't reproduce the error following the steps from this PR.

gedex commented:
> I couldn't test as I got this error and didn't understand the issue here:
Fatal error: Uncaught Error: Class 'WooCommerce'

Same as @claudiosanches, couldn't reproduce this. Make sure you ran `composer install` too.

claudiosanches committed:
Don't use WP_REST_Server::get_raw_data() inside REST API endpoints

claudiosanches reviewed (approved):
Looks good now. 
Just note that is missing schema, although isn't a public REST API, for now will be just fine without schema. Also this should require changes in the response, because we need fixed keys/indexes, but products objects returns IDs as keys.

@gedex Note that is required `-H "Content-Type: application/json"` header in order to make REST API properly read and decode JSON. So no more need for `WP_REST_Server::get_raw_data()`, see: 1f77efae4bbf8ed87c2e008c0209c28344b65932
